File: sdk/javascript/src/runtime-client.ts
import { requestJson, type HttpClientOptions } from "./http";

type RuntimeClientBaseOptions = {
  baseUrl: string;
  workspaceId: number;
  timeoutMs?: number;
};

export type EltRuntimeClientOptions = RuntimeClientBaseOptions & (
  | { projectKey: string; projectSecret: string; apiKey?: never }
  | { apiKey: string; projectKey?: never; projectSecret?: never }
);

export class EltRuntimeClient {
  private http: HttpClientOptions;
  private workspaceId: number;

  constructor(options: EltRuntimeClientOptions) {
    this.workspaceId = options.workspaceId;
    this.http = {
      baseUrl: options.baseUrl,
      auth: "projectKey" in options && options.projectKey !== undefined
        ? {
            type: "project",
            projectKey: options.projectKey,
            projectSecret: options.projectSecret,
          }
        : { type: "apiKey", apiKey: options.apiKey },
      timeoutMs: options.timeoutMs ?? 60_000,
    };
  }

  runPipeline(pipelineUuid: string, input?: Record<string, unknown>) {
    return requestJson<{ run_id: number; status: string }>(
      this.http,
      "POST",
      `/api/v1/runtime/workspaces/${this.workspaceId}/pipelines/${pipelineUuid}/run`,
      { input: input ?? null },
    );
  }

  runWorkflow(workflowUuid: string, input?: Record<string, unknown>) {
    return requestJson<{ run_id: number; status: string }>(
      this.http,
      "POST",
      `/api/v1/runtime/workspaces/${this.workspaceId}/workflows/${workflowUuid}/run`,
      { input: input ?? null },
    );
  }

  invokeRest(
    connectionId: number,
    body?: {
      path?: string;
      method?: string;
      variables?: Record<string, unknown>;
      body?: Record<string, unknown>;
    },
  ) {
    return requestJson<{
      status_code: number;
      data: unknown;
      headers?: Record<string, unknown>;
    }>(
      this.http,
      "POST",
      `/api/v1/runtime/workspaces/${this.workspaceId}/rest/${connectionId}/invoke`,
      body ?? {},
    );
  }

  /** Push a message (requires `queue:push` scope). */
  queuePush(queueName: string, payload?: Record<string, unknown>) {
    return requestJson<{ id: number; payload: unknown; created_at?: string }>(
      this.http,
      "POST",
      `/api/v1/runtime/workspaces/${this.workspaceId}/queues/${encodeURIComponent(queueName)}/push`,
      { payload: payload ?? {} },
    );
  }

  /** Pop oldest message — destructive; returns null when empty (requires `queue:pop` scope). */
  async queuePop(queueName: string) {
    const url = `/api/v1/runtime/workspaces/${this.workspaceId}/queues/${encodeURIComponent(queueName)}/pop`;
    const result = await requestJson<{ id: number; payload: unknown; created_at?: string } | undefined>(
      this.http,
      "POST",
      url,
    );
    return result ?? null;
  }

  /** Peek at the oldest message without removing it (requires `queue:read` scope). */
  async queuePeek(queueName: string) {
    const url = `/api/v1/runtime/workspaces/${this.workspaceId}/queues/${encodeURIComponent(queueName)}/peek`;
    const result = await requestJson<{ id: number; payload: unknown; created_at?: string } | undefined>(
      this.http,
      "GET",
      url,
    );
    return result ?? null;
  }

  /** Publish realtime notification (requires `notification:publish` scope). */
  notificationPublish(
    channel: string,
    payload?: Record<string, unknown>,
    target?: Record<string, unknown>,
  ) {
    return requestJson<{ channel: string; recipient_count: number; published: boolean }>(
      this.http,
      "POST",
      `/api/v1/runtime/workspaces/${this.workspaceId}/notifications/publish`,
      { channel, payload: payload ?? {}, ...(target ? { target } : {}) },
    );
  }
}

File: sdk/python/elt_sdk/runtime_client.py
"""Runtime SDK client — workspace API key authentication."""

from __future__ import annotations

from typing import Any, Dict, Optional

import httpx

from elt_sdk.client import EltClientError


class EltRuntimeClient:
    def __init__(
        self,
        base_url: str,
        *,
        api_key: str,
        workspace_id: int,
        timeout: float = 60.0,
    ):
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._workspace_id = workspace_id
        self._timeout = timeout

    def _headers(self) -> Dict[str, str]:
        return {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self._api_key}",
        }

    def _request(
        self,
        method: str,
        path: str,
        *,
        json: Optional[Dict[str, Any]] = None,
    ) -> Any:
        url = f"{self._base_url}{path}"
        with httpx.Client(timeout=self._timeout) as client:
            response = client.request(
                method,
                url,
                headers=self._headers(),
                json=json,
            )
        if response.status_code >= 400:
            raise EltClientError(response.text, response.status_code)
        if response.status_code == 204:
            return None
        return response.json()

    def run_pipeline(
        self,
        pipeline_uuid: str,
        *,
        input_payload: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        return self._request(
            "POST",
            f"/api/v1/runtime/workspaces/{self._workspace_id}/pipelines/{pipeline_uuid}/run",
            json={"input": input_payload},
        )

    def run_workflow(
        self,
        workflow_uuid: str,
        *,
        input_payload: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        return self._request(
            "POST",
            f"/api/v1/runtime/workspaces/{self._workspace_id}/workflows/{workflow_uuid}/run",
            json={"input": input_payload},
        )

    def invoke_rest(
        self,
        connection_id: int,
        *,
        path: Optional[str] = None,
        method: Optional[str] = None,
        variables: Optional[Dict[str, Any]] = None,
        body: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        return self._request(
            "POST",
            f"/api/v1/runtime/workspaces/{self._workspace_id}/rest/{connection_id}/invoke",
            json={
                "path": path,
                "method": method,
                "variables": variables or {},
                "body": body,
            },
        )

    def queue_push(
        self,
        queue_name: str,
        *,
        payload: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Push a message onto a queue. Requires scope queue:push."""
        return self._request(
            "POST",
            f"/api/v1/runtime/workspaces/{self._workspace_id}/queues/{queue_name}/push",
            json={"payload": payload or {}},
        )

    def queue_pop(self, queue_name: str) -> Optional[Dict[str, Any]]:
        """Pop the oldest message (destructive — message is removed). Requires scope queue:pop.

        Returns None if the queue is empty.
        """
        return self._request(
            "POST",
            f"/api/v1/runtime/workspaces/{self._workspace_id}/queues/{queue_name}/pop",
        )

    def notification_publish(
        self,
        channel: str,
        *,
        payload: Optional[Dict[str, Any]] = None,
        target: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Publish a realtime notification. Requires scope notification:publish."""
        body: Dict[str, Any] = {"channel": channel, "payload": payload or {}}
        if target is not None:
            body["target"] = target
        return self._request(
            "POST",
            f"/api/v1/runtime/workspaces/{self._workspace_id}/notifications/publish",
            json=body,
        )
